src/serverenvironment.cpp

```diff
@@ -1260,27 +1253,31 @@ void ServerEnvironment::step(float dtime)
 				abmhandler.apply(block);
 			}
 
-			u32 time_ms = timer.stop(true);
-			u32 max_time_ms = 200;
+			const u32 time_ms = timer.stop(true);
+			// allow up to 10% of the budget interval
+			const u32 max_time_ms = m_cache_abm_interval * 1000 / 10;
 			if (time_ms > max_time_ms) {
 				warningstream<<"active block modifiers took "
 					<<time_ms<<"ms (longer than "
 					<<max_time_ms<<"ms)"<<std::endl;
-				m_active_block_interval_overload_skip = (time_ms / max_time_ms) + 1;
+				m_active_block_interval_overload_skip = ((float)time_ms / max_time_ms);
+			} else {
+				m_active_block_interval_overload_skip = 1.0f;
 			}
 		}while(0);
 
 	/*
-		Step script environment (run global on_step())
+	  Step script environment (run global on_step())
 	*/
 	m_script->environment_Step(dtime);
 
 	/*
 		Step active objects
 	*/
-	{
+	if (m_active_object_interval.step(dtime, m_cache_ao_interval * m_active_object_interval_overload_skip)) {
+		float dtime = m_cache_ao_interval * m_active_object_interval_overload_skip;
```

Comment:
Why do you initialize `dtime` a second time here?

Reply:
Yeah. Was debating that too. It is the _actual_ dtime as far as the inside of the block is concerned.

There're other places in there where dtime is shadowed in a block. Happy to change this and those too.